### PR TITLE
DCOS-37613: enable Metronome actions to work with streams

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
**How to review**

Use the jobs tab, everything should work the same way as before

**Checklist**
- [X] Did you add a JIRA issue in a commit message or as part of the branch name?
- [X] Did you add new unit tests?
- [ ] Did you add new integration tests?
- [ ] If this is a regression, did you write a test to catch this in the future?

### CODE REVIEW

File: src/js/events/__tests__/MetronomeClient-test.ts [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+      MetronomeClient.fetchJobDetail(jobId);
+      expect(mockRequest).toHaveBeenCalled();
+    });
+
+    it("sends data to the correct URL", () => {
+      MetronomeClient.fetchJobDetail(jobId);
+      expect(mockRequest).toHaveBeenCalledWith(`${Config.metronomeAPI}/v1/jobs/${jobId}?embed=activeRuns&embed=schedules&embed=historySummary`);
+    });
+
+    it("emits the sucessful request result", marbles(function(m){
```
Comment (juliangieseke) [inline src/js/events/__tests__/MetronomeClient-test.ts]: 🔡 successful

File: src/js/utils/rxjsUtils.js [hunk 2]
```diff
@@ -8,6 +8,7 @@ export function linearBackoff(maxRetries = 5, delay = 1000, scheduler = null) {
   return errors =>
     errors
       .scan((count, error) => {
+        console.log("linearBackoff", count);
```
Comment (juliangieseke) [inline src/js/utils/rxjsUtils.js]: ⚠️

File: src/typings/dcos__http-service.d.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { Observable } from "rxjs";
+
+interface RequestOptions {
```
Comment (juliangieseke) [inline src/typings/dcos__http-service.d.ts]: IMHO we should add these to the package, its ours, so we can do it quick 🤔

File: src/js/events/MetronomeActions.js [hunk 4]
```diff
@@ -23,134 +21,93 @@ import {
 import AppDispatcher from "./AppDispatcher";
 import Config from "../config/Config";
 
+import {
+  createJob as createJobStream,
```
[the remark this answers is not in the document]
Comment (juliangieseke) [reply src/js/events/MetronomeActions.js]: Uh! Ok - _then_ I'd love to have the name reflect that (`…Factory` or something?)

File: src/js/events/MetronomeClient.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+export function fetchJobDetail(jobID: string): Observable<JobResponse> {
+  return request(`${Config.metronomeAPI}/v1/jobs/${jobID}?embed=activeRuns&embed=schedules&embed=historySummary`);
+}
+
+export function deleteJob(jobID: string, stopCurrentJobRuns = false): Observable<any> {
+  return request(`${Config.metronomeAPI}/v1/jobs/${jobID}?stopCurrentJobRuns=${stopCurrentJobRuns}`, { method: "DELETE" });
+}
+
+export function updateJob(jobID: string, data: JobData): Observable<JobResponse> {
+  return request(`${Config.metronomeAPI}/v0/scheduled-jobs/${jobID}`, {
```
[the remark this answers is not in the document]
Comment (juliangieseke) [reply src/js/events/MetronomeClient.ts]: does this mean that thee is _no_ way to archive this in `v1`? or is it just different? (then I'd say we should migrate)

File: src/js/events/__tests__/MetronomeActions-test.js [hunk 6]
```diff
@@ -1,40 +1,54 @@
...
 
-    it("dispatches the correct action when successful", function() {
+    it("dispatches the correct action when successful", function(done) {
       var id = AppDispatcher.register(function(payload) {
         var action = payload.action;
         AppDispatcher.unregister(id);
         expect(action.type).toEqual(
           ActionTypes.REQUEST_METRONOME_JOB_CREATE_SUCCESS
         );
+        done();
```
[the remark this answers is not in the document]
Comment (juliangieseke) [reply src/js/events/__tests__/MetronomeActions-test.js]: ok, let me rephrase my question: why is it _not_ needed in the other tests? :)

File: plugins/nodes/src/js/stores/__tests__/NodeHealthStore-test.js [hunk 7]
```diff
@@ -2,7 +2,7 @@ const RequestUtil = require("mesosphere-shared-reactjs").RequestUtil;
 
 const ActionTypes = require("#SRC/js/constants/ActionTypes");
 const AppDispatcher = require("#SRC/js/events/AppDispatcher");
-const Config = require("#SRC/js/config/Config");
+const Config = require("#SRC/js/config/Config").default;
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply plugins/nodes/src/js/stores/__tests__/NodeHealthStore-test.js]: 👍

File: src/js/config/Config.ts [hunk 8]
```diff
@@ -1,8 +1,43 @@
 /* eslint no-redeclare: 0 */
-import ConfigDev from "./Config.dev.js";
-import ConfigTest from "./Config.test.js";
 
-var Config = {
+import ConfigDev from "./Config.dev";
+import ConfigTest from "./Config.test";
+
+interface Configuration {
+  acsAPIPrefix: string
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply src/js/config/Config.ts]: Might be, would you mind filing an issue if you think this is important?

File: src/typings/dcos__http-service.d.ts (hunk 3 above)
Comment (DanielMSchmidt) [reply src/typings/dcos__http-service.d.ts]: Yes, we should. For me this is out of scope for this ticket.

File: plugins/services/src/js/containers/service-configuration/__tests__/ServiceConfigurationContainer-test.js [hunk 9]
```diff
@@ -5,6 +5,10 @@ const Application = require("../../../structs/Application");
 const ServiceConfigurationContainer = require("../ServiceConfigurationContainer");
 const ServiceConfigDisplay = require("../../../service-configuration/ServiceConfigDisplay");
 
+jest.mock("#SRC/js/events/MetronomeActions", () => ({
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply plugins/services/src/js/containers/service-configuration/__tests__/ServiceConfigurationContainer-test.js]: Because I would like to assert that the functions are called and not whatever they are doing in detail

**Comment by DanielMSchmidt** [comment]
@Fabs @juliangieseke I think I got it now. The implementation also lacked HTTP headers that were default previously and that are required, I just added them 👍

File: src/js/config/Config.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (juliangieseke) [reply src/js/config/Config.ts]: https://jira.mesosphere.com/browse/DCOS-37734

File: src/typings/dcos__http-service.d.ts (hunk 3 above)
Comment (Fabs) [reply src/typings/dcos__http-service.d.ts]: Think we should them create a ticket to move that for the package.

File: src/js/events/MetronomeActions.js [hunk 10]
```diff
@@ -22,201 +20,134 @@ import {
...
+import * as MetronomeClient from "./MetronomeClient";
 
 const MetronomeActions = {
   createJob(data) {
-    RequestUtil.json({
-      url: `${Config.metronomeAPI}/v0/scheduled-jobs`,
-      method: "POST",
-      data,
-      success() {
+    MetronomeClient.createJob(data).subscribe({
```
Comment (Fabs) [inline src/js/events/MetronomeActions.js]: Usually subscribe directly without handling the unsubscription is as bad as registering a callback without handling removing it. The reason is that infinite observers can cause a leak.

```js
const stream = myStream.subscribe(...);

// do the work

stream.unsubscribe();
```
I think here it is fine though because when the request completes, it is going to complete the observer (which does the unsubscribe). The problem I am still thinking here is that if the connection never finishes for some weird reason, `request` does not have a timeout. So  I wonder if we wanna add the timeout on http-service, or if we want have some type of timeout here.

https://github.com/dcos-labs/http-service/blob/89c781cfb54107ed45715d079c4b71149ee24b09/src/request.js

I think I need @orlandohohmeier opinion here.

Comment (Fabs) [reply src/js/events/MetronomeActions.js]: I talked with orlando. The connection has a timeout, which makes the request observable have one too. Which means that we do not care to unsubscribe. So this was never a problem.

**Comment by DanielMSchmidt** [comment]
@Fabs @juliangieseke I forgot to set some methods / data, should work now, at least it did on my machine :)

File: src/js/events/MetronomeClient.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    headers: defaultHeaders
+  });
+}
+
+export function runJob(jobID: string): Observable<any> {
+  return request(`${Config.metronomeAPI}/v1/jobs/${jobID}/runs`, { headers: defaultHeaders, method: "POST", body: "{}" });
+}
+
+export function stopJobRun(jobID: string, jobRunID: string): Observable<any> {
+  return request(`${Config.metronomeAPI}/v1/jobs/${jobID}/runs/${jobRunID}/actions/stop`, { headers: defaultHeaders, method: "POST", body: "{}" });
```
Comment (juliangieseke) [inline src/js/events/MetronomeClient.ts]: shouldnt this be prettified? 🤔

Comment (DanielMSchmidt) [reply src/js/events/MetronomeClient.ts]: Prettier doesn't seem to work currently on dcos-ui. I created a ticket for that: https://jira.mesosphere.com/browse/DCOS_OSS-3580

File: src/js/events/MetronomeClient.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+}
+interface IJobResponse {
+  id: string
+}
+interface IScheduleData {
+  id: string
+}
+
+const defaultHeaders = {
+  'Content-Type': "application/json; charset=utf-8"
```
Comment (nLight) [inline src/js/events/MetronomeClient.ts]: Why are we mixing single quotes with double quotes?

File: src/js/events/MetronomeClient.ts (hunk 11 above)
Comment (nLight) [reply src/js/events/MetronomeClient.ts]: What I did for extension framework is installing proper version and running prettier manually :shrug:

File: src/js/events/__tests__/MetronomeClient-test.ts [hunk 13]
```diff
@@ -0,0 +1,270 @@
+const mockRequest = jest.fn();
```
Comment (nLight) [inline src/js/events/__tests__/MetronomeClient-test.ts]: Jest usually hoists all its stuff so that we can have `import` at the top doesn't it?

File: .gitignore [hunk 14]
```diff
@@ -9,6 +9,7 @@ build/
 npm-debug.log
 node_modules/
 src/js/config/Config.dev.js
```
Comment (nLight) [inline .gitignore]: We don't really need both of them do we?

File: src/js/events/__tests__/MetronomeClient-test.ts (hunk 13 above)
Comment (DanielMSchmidt) [reply src/js/events/__tests__/MetronomeClient-test.ts]: Yes and no, it hoists the `jest.mock` call, but not the constant, therefore I get this:


```log
 FAIL  src/js/events/__tests__/MetronomeClient-test.ts
  ● Test suite failed to run

    ReferenceError: mockRequest is not defined

      4 | import { Observable } from "rxjs/Observable";
      5 |
    > 6 | import Config from "../config/Config";
```

File: .gitignore (hunk 14 above)
Comment (DanielMSchmidt) [reply .gitignore]: I thought it would be nicer for people migrating over (got a lot of bad vibes for not ignoring `jest/config.json` although we didn't need it anymore)

File: src/js/events/MetronomeClient.ts (hunk 11 above)
Comment (nLight) [reply src/js/events/MetronomeClient.ts]: Did you run prettier eventually?

File: src/js/events/MetronomeActions.js [hunk 15]
```diff
@@ -22,201 +20,134 @@ import {
...
-    RequestUtil.json({
-      url: `${Config.metronomeAPI}/v1/jobs/${jobID}`,
-      data: [
-        { name: "embed", value: "activeRuns" },
-        { name: "embed", value: "history" },
-        { name: "embed", value: "schedules" }
-      ],
-      success(response) {
+    MetronomeClient.fetchJobDetail(jobID).subscribe({
+      next: response =>
```
Comment (Fabs) [inline src/js/events/MetronomeActions.js]: This bit is missing this data part, I guess. I am saying that because I still see a bug on testing the jobs between nuc and this PR, and it is not showing all the previous runs.
```
     data: [	
         { name: "embed", value: "activeRuns" },	
          { name: "embed", value: "history" },
         { name: "embed", value: "schedules" }	
     ],
```

File: src/js/events/MetronomeActions.js [hunk 16]
```diff
@@ -22,201 +20,134 @@ import {
...
-              type: REQUEST_METRONOME_JOBS_ONGOING
-            });
-          }
-        });
-      };
-    },
-    { delayAfterCount: Config.delayAfterErrorCount }
-  ),
+  fetchJobs() {
+    MetronomeClient.fetchJobs().subscribe({
```
Comment (Fabs) [inline src/js/events/MetronomeActions.js]: Think also here we lose the data we want to embed so we get what we need

Comment (Fabs) [reply src/js/events/MetronomeActions.js]:
```
     data: [	
         { name: "embed", value: "activeRuns" },	
          { name: "embed", value: "history" },
         { name: "embed", value: "schedules" }	
     ],
```

File: src/js/events/__tests__/MetronomeActions-test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply src/js/events/__tests__/MetronomeActions-test.js]: yeah, they weren't even executing 🙈

File: src/js/events/MetronomeActions.js (hunk 15 above)
Comment (DanielMSchmidt) [reply src/js/events/MetronomeActions.js]: You are right, fixing this right now 👍

**Comment by DanielMSchmidt** [comment]
@Fabs @juliangieseke @nLight Sorry for letting you run so many cycles, I will deliver better work next time.

File: src/js/config/Config.ts [hunk 17]
```diff
@@ -3,7 +3,7 @@
 import ConfigDev from "./Config.dev";
 import ConfigTest from "./Config.test";
 
-interface Configuration {
+interface IConfiguration {
```
Comment (juliangieseke) [inline src/js/config/Config.ts]: Uh, didnt we say that we dont want to do it like this? 🤔

File: src/js/config/Config.ts [hunk 18]
```diff
@@ -1,8 +1,45 @@
 /* eslint no-redeclare: 0 */
-import ConfigDev from "./Config.dev.js";
-import ConfigTest from "./Config.test.js";
 
-var Config = {
+import ConfigDev from "./Config.dev";
+import ConfigTest from "./Config.test";
+
+interface IConfiguration {
```
Comment (juliangieseke) [inline src/js/config/Config.ts]: Uh, didnt we say that we dont want to do it like this? 🤔

**Comment by juliangieseke** [comment]
CI is legitimately red :(

File: src/js/config/Config.ts (hunk 18 above)
Comment (DanielMSchmidt) [reply src/js/config/Config.ts]: We still have it in our tslint config

**Comment by Fabs** [comment]
Manually testing looks fine, looks like there are some CI problems. It was a journey but we got there. Congrats @DanielMSchmidt :-).
